### PR TITLE
Add manifest V3 true HMR

### LINKED ISSUE #8168: v2.6.0 BUG: web-extension, content script, importScripts() is not supported in module workers
# 🐛 bug report

There is a regression with how Parcel 2.6.0 HMR handles content script as well as background script for web extension runtime. I got an error (pasted below) when trying to use `@parcel/config-webextension@2.6.0` to build an extension with just a single content script.

## 🎛 Configuration (.babelrc, package.json, cli command)

The content script looks like this:

```
window.addEventListener("load", () => {
  console.log("content script loaded")

  document.body.style.background = "pink"
})
```

## 🤔 Expected Behavior

In 2.5.0, the above content script worked. 

## 😯 Current Behavior

Upon upgrading to 2.6.0, it threw the error below:

```
[Error: ENOENT: no such file or directory, open 'P:\Projects\plasmo-corp\plasmo\node_modules\.pnpm\@parcel+runtime-browser-hmr@2.6.0_@parcel+core@2.6.0\node_modules\@parcel\runtime-browser-hmr\lib\runtime-62b1b20e127e8e50.js'] {
  errno: -4058,
  code: 'ENOENT',
  syscall: 'open',
  path: 'P:\\Projects\\plasmo-corp\\plasmo\\node_modules\\.pnpm\\@parcel+runtime-browser-hmr@2.6.0_@parcel+core@2.6.0\\node_modules\\@parcel\\runtime-browser-hmr\\lib\\runtime-62b1b20e127e8e50.js'
}
           importScripts() is not supported in module workers.
           Use a static `import`, or dynamic `import()` instead.
           P:\Projects\plasmo-corp\plasmo\node_modules\.pnpm\@parcel+runtime-browser-hmr@2.6.0_@parcel+core@2.6.0\node_modules\@parcel\runtime-browser-hmr\lib\runtime-62b1b20e127e8e50.js:320:17
```

## 💁 Possible Solution

We are investigating the HMR runtime #8159. Will continue doing so for a bit more, haven't figured out what's causing this issue but I suspect contentscripts and background scripts are using the wrong transformer/resolver?

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Parcel           | 2.6.0
| Node             | 16.14.0
| npm/Yarn         | 8.3.1
| Operating System | Windows

### PR DESCRIPTION
**Pull request by 101arrowz**
# ↪️ Pull Request

Adds true HMR for all assets in a Manifest V3 web extension. Unfortunately this is mostly useless for the foreseeable future because of [a Chromium bug](https://bugs.chromium.org/p/chromium/issues/detail?id=1255412&q=localhost%20manifest%20v3&can=2#c12). However hot reload at least works for the options page and sandbox pages now, just not for content scripts.

Fixes #8168, which currently breaks pretty much all MV3 extensions with the default background script.

## ✔️ PR Todo

- [ ] Added/updated unit tests for this change
- [ ] Filled out test instructions (In case there aren't any unit tests)
- [ ] Included links to related issues/PRs

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 8.73s | -25.00ms |
| Cached | 464.00ms | +27.00ms ⚠️ |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 1.55m | -926.00ms |
| Cached | 2.46s | -49.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.41s | +71.00ms |
| Cached | 254.00ms | -3.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

*No bundle changes detected.*

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/334308822511256146)**

**Comment by 101arrowz** [comment]
This is very imperfect but fixes a major bug with `@parcel/config-webextension` on v2.6.0 caused by a mistake in the HMR runtime. If nothing else, it should be merged to fix that issue.

File: packages/runtimes/hmr/src/loaders/hmr-runtime.js [hunk 1]
```diff
@@ -325,8 +345,37 @@ function reloadCSS() {
...
+        script.onerror = reject;
+        document.head?.appendChild(script);
+      });
+    } else if (typeof importScripts === 'function') {
+      // Worker scripts
+      return new Promise((resolve, reject) => {
+        try {
+          __parcel__importScripts__(asset.url);
+          resolve();
+        } catch (err) {
```
Comment (devongovett) [inline packages/runtimes/hmr/src/loaders/hmr-runtime.js]: `importScripts` is synchronous, so this would currently catch any type error thrown by the imported code too. Maybe better to detect ESM (e.g. try..catch around `import.meta`, which throws an error when not in a module).

File: packages/runtimes/hmr/src/loaders/hmr-runtime.js [hunk 2]
```diff
@@ -93,6 +99,25 @@ if ((!parent || !parent.isParcelRequire) && typeof WebSocket !== 'undefined') {
...
+  var globalObject =
+    typeof globalThis !== 'undefined'
+      ? globalThis
+      : typeof self !== 'undefined'
+      ? self
+      : typeof window !== 'undefined'
+      ? window
+      : typeof global !== 'undefined'
+      ? global
+      : {};
```
Comment (devongovett) [inline packages/runtimes/hmr/src/loaders/hmr-runtime.js]: Why was this necessary? `global` should be injected by parcel when needed...

File: packages/runtimes/hmr/src/loaders/hmr-runtime.js [hunk 3]
```diff
@@ -325,8 +345,37 @@ function reloadCSS() {
...
+      return new Promise((resolve, reject) => {
+        script.onload = () => resolve(script);
+        script.onerror = reject;
+        document.head?.appendChild(script);
+      });
+    } else if (typeof importScripts === 'function') {
+      // Worker scripts
+      return new Promise((resolve, reject) => {
+        try {
+          __parcel__importScripts__(asset.url);
```
Comment (devongovett) [inline packages/runtimes/hmr/src/loaders/hmr-runtime.js]: thanks for fixing this.

File: packages/reporters/dev-server/src/Server.js [hunk 4]
```diff
@@ -369,19 +353,15 @@ export default class Server {
 
   sendError(res: Response, statusCode: number) {
     res.statusCode = statusCode;
-    setHeaders(res);
```
Comment (devongovett) [inline packages/reporters/dev-server/src/Server.js:372]: why were the headers removed from a bunch of places?

**Review by devongovett — COMMENTED** [review]
Thanks, this looks good. A few questions.

File: packages/reporters/dev-server/src/Server.js (hunk 4 above)
Comment (101arrowz) [reply packages/reporters/dev-server/src/Server.js:372]: `setHeaders` is already called for any request due to the initial middleware so it's redundant.

File: packages/runtimes/hmr/src/loaders/hmr-runtime.js (hunk 2 above)
Comment (101arrowz) [reply packages/runtimes/hmr/src/loaders/hmr-runtime.js]: It wasn't for some reason. I don't think `global` is actually a property that exists in browsers anyway...just in Node.js

File: packages/runtimes/hmr/src/loaders/hmr-runtime.js (hunk 1 above)
Comment (101arrowz) [reply packages/runtimes/hmr/src/loaders/hmr-runtime.js]: I thought `import.meta` was invalid syntax in some browsers but I can change it.

File: packages/transformers/webextension/src/schema.js [hunk 5]
```diff
@@ -449,6 +448,13 @@ export const MV3Schema = ({
       additionalProperties: false,
     },
     host_permissions: arrStr,
+    sandbox: {
```
Comment (louisgv) [inline packages/transformers/webextension/src/schema.js:451]: Duplicate property?

File: packages/runtimes/hmr/src/loaders/hmr-runtime.js (hunk 2 above)
Comment (devongovett) [reply packages/runtimes/hmr/src/loaders/hmr-runtime.js]: Parcel should be polyfilling it https://parceljs.org/features/node-emulation/#shimming-builtin-node-globals

File: packages/transformers/webextension/src/schema.js (hunk 5 above)
Comment (devongovett) [reply packages/transformers/webextension/src/schema.js:451]: There are two different schemas in this file, so they are separate.

File: packages/runtimes/hmr/src/loaders/hmr-runtime.js (hunk 1 above)
Comment (devongovett) [reply packages/runtimes/hmr/src/loaders/hmr-runtime.js]: Going to send the outputFormat from the HMR server as part of the asset instead.

File: packages/runtimes/hmr/src/loaders/hmr-runtime.js (hunk 2 above)
Comment (devongovett) [reply packages/runtimes/hmr/src/loaders/hmr-runtime.js]: I tested and it seems to be working, but maybe I missed something

**Comment by devongovett** [comment]
Pushed a few small updates. Hopefully I didn't break anything. Let me know.

**Review by devongovett — APPROVED** [review]
